internal/provider/user_data_source.go

```diff
@@ -91,46 +93,123 @@ func (d *userDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, r
 // Read refreshes the Terraform state with the latest data.
 func (d *userDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
 	tflog.Debug(ctx, "Preparing to read user data source")
-	var state userDataSourceModel
+	var config userDataSourceModel
 
-	resp.Diagnostics.Append(req.Config.Get(ctx, &state)...)
+	resp.Diagnostics.Append(req.Config.Get(ctx, &config)...) // capture user input
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	userId, err := strconv.Atoi(state.Id.ValueString())
-	if err != nil {
+	if config.Id.IsUnknown() {
+		resp.Diagnostics.AddError("Cannot use unknown value for id", "Provide a concrete id or omit it in favour of the email lookup.")
+		return
+	}
+	if config.Email.IsUnknown() {
+		resp.Diagnostics.AddError("Cannot use unknown value for email", "Provide a concrete email or omit it in favour of the id lookup.")
+		return
+	}
+
+	idProvided := !config.Id.IsNull() && config.Id.ValueString() != ""
+	emailProvided := !config.Email.IsNull() && config.Email.ValueString() != ""
+
+	if !idProvided && !emailProvided {
 		resp.Diagnostics.AddError(
-			fmt.Sprintf("User id was not a number %s", state.Id.ValueString()),
-			err.Error(),
+			"Missing lookup attribute",
+			"Specify either the user's id or email to retrieve the user.",
 		)
 		return
 	}
 
-	user, api_response, err := d.client.UsersAPI.GetUser(ctx, int32(userId)).Execute()
-	if !ValidateApiResponse(api_response, 200, &resp.Diagnostics, err) {
-		return
+	var user *unleash.UserSchema
+
+	if idProvided {
+		idValue, err := strconv.Atoi(config.Id.ValueString())
+		if err != nil {
+			resp.Diagnostics.AddError(
+				fmt.Sprintf("User id was not a number %s", config.Id.ValueString()),
+				err.Error(),
+			)
+			return
+		}
+
+		apiUser, apiResponse, err := d.client.UsersAPI.GetUser(ctx, int32(idValue)).Execute()
+		if !ValidateApiResponse(apiResponse, 200, &resp.Diagnostics, err) {
+			return
+		}
+		user = apiUser
+
+		if emailProvided {
```

Comment:
Struggling a bit with the nesting here. Not sure but I think flattening this if statement may help to make this monster a bit more readble. Tempted to say the conditions for the if statement can just be a single well named function because I also can't spot at a glance what this case represents